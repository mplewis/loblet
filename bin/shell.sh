#!/bin/bash
set -euo pipefail
IFS=$'\n\t'

bin/build.sh
docker run --entrypoint=bash -it loblet
