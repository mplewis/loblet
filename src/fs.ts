import e from 'expand-tilde'
import { readFile } from 'fs-extra'

export async function read (path: string): Promise<string> {
  return (await readFile(e(path))).toString()
}
