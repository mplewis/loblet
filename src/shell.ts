import { exec as _exec } from 'shelljs'
import { red, blue } from 'chalk'

interface Options {
  echo?: boolean;
  exitCode?: number;
}

const DEFAULTS: Options = {
  echo: false,
  exitCode: 0
}

const COLORS = {
  ECHO: blue,
  ERROR: red
}

export async function exec (cmd: string, options?: Options): Promise<{ stdout: string; stderr: string }> {
  const resolvedOptions: Options = Object.assign({}, DEFAULTS, options)
  const silent = !resolvedOptions.echo
  const expectedExitCode = resolvedOptions.exitCode

  const log = (msg: string): void => { if (!silent) console.log(msg) }

  log(COLORS.ECHO(cmd))
  const process = await _exec(cmd, { async: true, silent })
  return new Promise((resolve, reject) => {
    let stdout = ''
    let stderr = ''
    process.stdout.on('data', s => { stdout += s })
    process.stderr.on('data', s => { stderr += s })
    process.on('close', (code) => {
      if (code === expectedExitCode) return resolve({ stdout, stderr })

      const msg = `Process exited with code ${code}: ${cmd}`
      log(COLORS.ERROR(msg))
      if (stdout.length > 0) {
        log(COLORS.ERROR('stdout:'))
        log(stdout)
      }
      if (stderr.length > 0) {
        log(COLORS.ERROR('stderr:'))
        log(stderr)
      }
      return reject(new Error(msg))
    })
  })
}
