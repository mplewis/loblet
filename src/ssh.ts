import { read } from './fs'
import { ensureFile, writeFile } from 'fs-extra'
import e from 'expand-tilde'

export const PRIVATE_KEY_FILE = e('~/.ssh/id_rsa')
export const KNOWN_HOSTS_FILE = e('~/.ssh/known_hosts')

export async function setPrivKey (privKey: string): Promise<void> {
  await ensureFile(PRIVATE_KEY_FILE)
  await writeFile(PRIVATE_KEY_FILE, privKey, { mode: 0o644 })
}

export async function ensureKnownHost (pubKeyLine: string): Promise<void> {
  await ensureFile(KNOWN_HOSTS_FILE)
  const lines = (await read(KNOWN_HOSTS_FILE)).split('\n')
  if (lines.includes(pubKeyLine)) return

  lines.push(pubKeyLine)
  await writeFile(KNOWN_HOSTS_FILE, lines.join('\n'))
}
