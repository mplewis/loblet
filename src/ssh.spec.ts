import { ensureKnownHost, KNOWN_HOSTS_FILE } from './ssh'
import ensureContainerized from '../spec/utils/ensure_containerized'
import { privKey } from '../spec/fixtures/ssh'
import { unlink, ensureFile, writeFile } from 'fs-extra'
import { read } from './fs'
import { exec } from './shell'
import expandTilde = require('expand-tilde')

ensureContainerized()

const ID_RSA = expandTilde('~/.ssh/id_rsa')
const GITHUB_KNOWN_HOST = 'github.com ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEAq2A7hRGmdnm9tUDbO9IDSwBK6TbQa+PXYPCPy6rbTrTtw7PHkccKrpp0yVhp5HdEIcKr6pLlVDBfOLX9QUsyCOV0wzfjIJNlGEYsdlLJizHhbn2mUjvSAHQqZETYP81eFzLQNnPHt4EVVUh7VfDESU84KezmD5QlWpXLmvU31/yMf+Se8xhHTvKSCZIFImWwoG6mbUoWf9nzpIoaSjB+weqqUUmpaaasXVal72J+UX2B+2RPW3RcT0eOzQgqlJL3RKrTJvdsjE3JEAvGq3lGHSZXy28G3skua2SmVi/w4yCE6gbODqnTWlg7+wC604ydGXA8VJiS5ap43JXiUFFAaQ=='

async function itPopulatesTheFileAsExpected (times = 1): Promise<void> {
  it('populates the file as expected', async () => {
    for (let i = 0; i < times; i++) {
      await ensureKnownHost(GITHUB_KNOWN_HOST)
    }
    expect(await read(KNOWN_HOSTS_FILE)).toMatchSnapshot()
  })
}

describe('ensureKnownHost', () => {
  afterEach(async () => { await unlink(KNOWN_HOSTS_FILE) })

  describe('when no file exists', () => {
    itPopulatesTheFileAsExpected()
  })

  describe('when an empty file exists', () => {
    beforeEach(async () => { await ensureFile(KNOWN_HOSTS_FILE) })
    itPopulatesTheFileAsExpected()
  })

  describe('when the file exists with no matching line', () => {
    beforeEach(async () => {
      const lines = [
        'mplewis.com ssh-rsa SOME-FAKE-KEY-DATA',
        'fdnt.me ssh-rsa SOME-FAKE-KEY-DATA'
      ]
      await writeFile(KNOWN_HOSTS_FILE, lines.join('\n'))
    })
    itPopulatesTheFileAsExpected()
  })

  describe('ensuring the line exists multiple times in a row', () => {
    beforeEach(async () => {
      const lines = [
        'mplewis.com ssh-rsa SOME-FAKE-KEY-DATA',
        'fdnt.me ssh-rsa SOME-FAKE-KEY-DATA'
      ]
      await writeFile(KNOWN_HOSTS_FILE, lines.join('\n'))
    })
    itPopulatesTheFileAsExpected(5)
  })

  describe('when the file exists with a matching line', () => {
    beforeEach(async () => {
      const lines = [
        'mplewis.com ssh-rsa SOME-FAKE-KEY-DATA',
        'fdnt.me ssh-rsa SOME-FAKE-KEY-DATA',
        GITHUB_KNOWN_HOST
      ]
      await writeFile(KNOWN_HOSTS_FILE, lines.join('\n'))
    })
    itPopulatesTheFileAsExpected()
  })

  describe('after ensuring the GitHub known host', () => {
    beforeEach(async () => {
      await ensureKnownHost(GITHUB_KNOWN_HOST)
      await writeFile(ID_RSA, privKey, { mode: 0o600 })
    })
    afterEach(async () => { await unlink(ID_RSA) })
    it('connects to GitHub successfully', async () => {
      const { stderr } = await exec('ssh -T git@github.com', { exitCode: 1 })
      expect(stderr).toContain("Hi mplewis/private-clonable-repo! You've successfully authenticated, but GitHub does not provide shell access.")
    })
  })
})
