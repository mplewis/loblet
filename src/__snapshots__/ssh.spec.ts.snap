// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ensureKnownHost ensuring the line exists multiple times in a row populates the file as expected 1`] = `
"mplewis.com ssh-rsa SOME-FAKE-KEY-DATA
fdnt.me ssh-rsa SOME-FAKE-KEY-DATA
github.com ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEAq2A7hRGmdnm9tUDbO9IDSwBK6TbQa+PXYPCPy6rbTrTtw7PHkccKrpp0yVhp5HdEIcKr6pLlVDBfOLX9QUsyCOV0wzfjIJNlGEYsdlLJizHhbn2mUjvSAHQqZETYP81eFzLQNnPHt4EVVUh7VfDESU84KezmD5QlWpXLmvU31/yMf+Se8xhHTvKSCZIFImWwoG6mbUoWf9nzpIoaSjB+weqqUUmpaaasXVal72J+UX2B+2RPW3RcT0eOzQgqlJL3RKrTJvdsjE3JEAvGq3lGHSZXy28G3skua2SmVi/w4yCE6gbODqnTWlg7+wC604ydGXA8VJiS5ap43JXiUFFAaQ=="
`;

exports[`ensureKnownHost when an empty file exists populates the file as expected 1`] = `
"
github.com ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEAq2A7hRGmdnm9tUDbO9IDSwBK6TbQa+PXYPCPy6rbTrTtw7PHkccKrpp0yVhp5HdEIcKr6pLlVDBfOLX9QUsyCOV0wzfjIJNlGEYsdlLJizHhbn2mUjvSAHQqZETYP81eFzLQNnPHt4EVVUh7VfDESU84KezmD5QlWpXLmvU31/yMf+Se8xhHTvKSCZIFImWwoG6mbUoWf9nzpIoaSjB+weqqUUmpaaasXVal72J+UX2B+2RPW3RcT0eOzQgqlJL3RKrTJvdsjE3JEAvGq3lGHSZXy28G3skua2SmVi/w4yCE6gbODqnTWlg7+wC604ydGXA8VJiS5ap43JXiUFFAaQ=="
`;

exports[`ensureKnownHost when no file exists populates the file as expected 1`] = `
"
github.com ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEAq2A7hRGmdnm9tUDbO9IDSwBK6TbQa+PXYPCPy6rbTrTtw7PHkccKrpp0yVhp5HdEIcKr6pLlVDBfOLX9QUsyCOV0wzfjIJNlGEYsdlLJizHhbn2mUjvSAHQqZETYP81eFzLQNnPHt4EVVUh7VfDESU84KezmD5QlWpXLmvU31/yMf+Se8xhHTvKSCZIFImWwoG6mbUoWf9nzpIoaSjB+weqqUUmpaaasXVal72J+UX2B+2RPW3RcT0eOzQgqlJL3RKrTJvdsjE3JEAvGq3lGHSZXy28G3skua2SmVi/w4yCE6gbODqnTWlg7+wC604ydGXA8VJiS5ap43JXiUFFAaQ=="
`;

exports[`ensureKnownHost when the file exists with a matching line populates the file as expected 1`] = `
"mplewis.com ssh-rsa SOME-FAKE-KEY-DATA
fdnt.me ssh-rsa SOME-FAKE-KEY-DATA
github.com ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEAq2A7hRGmdnm9tUDbO9IDSwBK6TbQa+PXYPCPy6rbTrTtw7PHkccKrpp0yVhp5HdEIcKr6pLlVDBfOLX9QUsyCOV0wzfjIJNlGEYsdlLJizHhbn2mUjvSAHQqZETYP81eFzLQNnPHt4EVVUh7VfDESU84KezmD5QlWpXLmvU31/yMf+Se8xhHTvKSCZIFImWwoG6mbUoWf9nzpIoaSjB+weqqUUmpaaasXVal72J+UX2B+2RPW3RcT0eOzQgqlJL3RKrTJvdsjE3JEAvGq3lGHSZXy28G3skua2SmVi/w4yCE6gbODqnTWlg7+wC604ydGXA8VJiS5ap43JXiUFFAaQ=="
`;

exports[`ensureKnownHost when the file exists with no matching line populates the file as expected 1`] = `
"mplewis.com ssh-rsa SOME-FAKE-KEY-DATA
fdnt.me ssh-rsa SOME-FAKE-KEY-DATA
github.com ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEAq2A7hRGmdnm9tUDbO9IDSwBK6TbQa+PXYPCPy6rbTrTtw7PHkccKrpp0yVhp5HdEIcKr6pLlVDBfOLX9QUsyCOV0wzfjIJNlGEYsdlLJizHhbn2mUjvSAHQqZETYP81eFzLQNnPHt4EVVUh7VfDESU84KezmD5QlWpXLmvU31/yMf+Se8xhHTvKSCZIFImWwoG6mbUoWf9nzpIoaSjB+weqqUUmpaaasXVal72J+UX2B+2RPW3RcT0eOzQgqlJL3RKrTJvdsjE3JEAvGq3lGHSZXy28G3skua2SmVi/w4yCE6gbODqnTWlg7+wC604ydGXA8VJiS5ap43JXiUFFAaQ=="
`;
